File: examples/ch06/after.py
"""Chapter 6 · AFTER: a cart that keeps its own promises.

A narrow public surface over a hidden implementation. `total` is computed, so
it can never go stale or lie; `discount_rate` is validated; the item list is
hidden behind a read-only view. There is no way to put this cart in an illegal
state from the outside.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class LineItem:
    name: str
    price: float


class ShoppingCart:
    def __init__(self) -> None:
        self._items: list[LineItem] = []
        self._discount_rate: float = 0.0

    def add_item(self, item: LineItem) -> None:
        self._items.append(item)

    @property
    def items(self) -> tuple[LineItem, ...]:
        return tuple(self._items)  # a read-only view; mutating it can't corrupt us

    @property
    def discount_rate(self) -> float:
        return self._discount_rate

    @discount_rate.setter
    def discount_rate(self, rate: float) -> None:
        if not 0.0 <= rate <= 1.0:
            raise ValueError(f"discount_rate must be in [0, 1], got {rate}")
        self._discount_rate = rate

    @property
    def total(self) -> float:
        subtotal = sum(item.price for item in self._items)
        return round(subtotal * (1 - self._discount_rate), 2)

File: examples/ch06/ts/after.ts
// Chapter 6 · AFTER: a cart that keeps its own promises.
//
// A narrow public surface over a hidden implementation. `#items` and
// `#discountRate` are truly private (no caller can reach them). `total` is a
// computed getter, so it can never go stale or lie; `discountRate` validates on
// the way in; `items` returns a readonly view. There is no way to put this cart
// into an illegal state from the outside.

import { type LineItem } from "./models";

export class ShoppingCart {
  #items: LineItem[] = [];
  #discountRate = 0;

  addItem(item: LineItem): void {
    this.#items.push(item);
  }

  get items(): readonly LineItem[] {
    return [...this.#items]; // a readonly copy; mutating it can't corrupt us
  }

  get discountRate(): number {
    return this.#discountRate;
  }

  set discountRate(rate: number) {
    if (rate < 0 || rate > 1) {
      throw new RangeError(`discountRate must be in [0, 1], got ${rate}`);
    }
    this.#discountRate = rate;
  }

  get total(): number {
    // Computed on read — never stale. No setter, so the lie is unrepresentable.
    const subtotal = this.#items.reduce((sum, item) => sum + item.price, 0);
    return Math.round(subtotal * (1 - this.#discountRate) * 100) / 100;
  }
}
